--- server/main.c	2020-09-02 09:24:31.000000000 -0600
+++ server/main.AFL++.c	2020-09-04 14:28:24.000000000 -0600
@@ -14,6 +14,43 @@
  * limitations under the License.
  */
 
+// AFL++ SNIP
+#include <arpa/inet.h>
+#include <errno.h>
+#include <linux/sched.h>
+#include <net/if.h>
+#include <net/route.h>
+#include <netinet/ip6.h>
+#include <netinet/tcp.h>
+#include <sched.h>
+#include <signal.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h>
+#include <sys/ioctl.h>
+#include <sys/resource.h>
+#include <sys/socket.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#ifndef __AFL_FUZZ_TESTCASE_LEN
+  unsigned int  fuzz_len;
+  unsigned char fuzz_buf[1024000];
+
+  #define __AFL_FUZZ_TESTCASE_LEN fuzz_len
+  #define __AFL_FUZZ_TESTCASE_BUF fuzz_buf
+  #define __AFL_FUZZ_INIT() void sync(void);
+  #define __AFL_LOOP(x) ((fuzz_len = read(0, fuzz_buf, sizeof(fuzz_buf))) > 0 ?
+  #define __AFL_INIT() sync()
+
+#endif
+
+__AFL_FUZZ_INIT();
+// AFL++ EOS
+
 #include "apr.h"
 #include "apr_strings.h"
 #include "apr_getopt.h"
@@ -482,8 +519,135 @@
     ap_log_error(APLOG_MARK, APLOG_STARTUP, 0, NULL,
                  "  -X                 : debug mode (only one worker, do not detach)");
 
-    destroy_and_exit_process(process, 1);
+// AFL++ SNIP
+    destroy_and_exit_process(process, 0);
+// AFL++ EOS
+}
+
+// AFL++ SNIP
+static void INIT_AFL_MUTATION_SESSION(process_rec *process)
+{
+  // DEFINE __AFL_LOOP() MUTATION ITERATIONS
+  // This may be waaaaay too low but hey let's give it a go...
+  //int iterations = 10;
+  // Recommended Default (At least initially)
+  int iterations = 1000;
+  //int iterations = 3000;
+  // This may be waaaaay too high but hey let's give it a go...
+  //int iterations = 300000;
+  // This is the max number of mutation iterations 
+  // may be used / loop (if 100% stability is reached)
+  //int iterations = 2147483647;
+
+  // Init Mutated Payload Vars
+  unsigned int len;
+  unsigned char *buf;
+
+  printf("[+]Deferring Initialization for Speed Boost...\n");
+  __AFL_INIT();
+  buf = __AFL_FUZZ_TESTCASE_BUF;
+  printf("[+] ENTERING __AFL_LOOP(%d)...\n", iterations);
+  while ( __AFL_LOOP(iterations) ){
+    len = __AFL_FUZZ_TESTCASE_LEN;
+    //if (len == 0) continue;
+    printf("before buffer reset:\n");
+    printf("input: %zd \"%s\"\n", len, buf);
+
+    printf("=> Resetting buffer to %d\n", len);
+    memset(buf, 0, len);
+
+    printf("after buffer reset:\n");
+    printf("input: %zd \"%s\"\n", len, buf);
+
+    printf("=> reading next mutation...\n");
+    unsigned int read_bytes = read(0, buf, len);
+
+    // Prevents mutated test cases
+    // from being submitted to httpd?
+    buf[len-2] = '\r';
+    buf[len-1] = '\n';
+    buf[len] = '\0';
+
+    printf("=> init sockfd\n");
+    int sockfd = socket(AF_INET, SOCK_STREAM, IPPROTO_IP);
+    if (sockfd == -1) {
+      printf("[!] Ugh, sockfd error!\n");
+      perror("socket");
+      _exit(1);
+    }
+
+    printf("=> init setsockopt\n");
+    int sz = (1024 * 1024);
+    if (setsockopt(sockfd, SOL_SOCKET, SO_SNDBUF, &sz, sizeof(sz)) == -1) {
+      printf("[!] Ugh, setsockopt error!\n");
+      perror("setsockopt");
+      _exit(1);
+    }
+
+    printf("=> connecting to tcp port 80\n", buf);
+    struct sockaddr_in saddr;
+    saddr.sin_family = AF_INET;
+    saddr.sin_port = htons(80);
+    saddr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+    if (connect(sockfd, &saddr, sizeof(saddr)) == -1) {
+      printf("[!] Ugh, failed to connect do sockfd!\n");
+      continue;
+    }
+
+    printf("=> sending buf %s\n", buf);
+    if (send(sockfd, buf, read_bytes, MSG_NOSIGNAL) != read_bytes) {
+      printf("[!] Ugh, failed to send payload over sockfd!\n");
+      _exit(1);
+    }
+
+    printf("=> buf sent %s\n", &buf);
+
+    if (shutdown(sockfd, SHUT_WR) == -1) {
+      perror("[!] Ugh, failed to shutdown sockfd!\n");
+      _exit(1);
+    }
+
+    char b[1024 * 1024];
+    while (recv(sockfd, b, sizeof(b), MSG_WAITALL) > 0) ;
+      printf("=> Received from sockfd %s\n", b);
+
+      printf("=> Closing sockfd...\n");
+      close(sockfd);
+      printf("=> successfully closed sockfd\n");
+  }
+  printf("[+] EXITING __AFL_LOOP(%d)\n", iterations);
+
+  //printf("Removing %s\n", DEFAULT_PIDLOG);
+  //printf("HTTPD_ROOT %s\n", HTTPD_ROOT);
+  int del_pidlog;
+  del_pidlog = remove("/fuzz_session/httpd/logs/httpd.pid");
+  if ( del_pidlog != 0 ) {
+    printf("[!] Ugh, unable to remove %s", DEFAULT_PIDLOG);
+    _exit(1);
+  } else {
+    printf("[+] REMOVED %s\n", DEFAULT_PIDLOG);
+    _exit(0);
+  }
+}
+
+static void INIT_AFL_THREAD(process_rec *process)
+{
+    pthread_t t;
+    pthread_attr_t attr;
+
+    pthread_attr_init(&attr);
+    pthread_attr_setstacksize(&attr, 1024 * 1024 * 8);
+    pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);
+
+    printf("[*]: CREATING AFL++ mutation thread...\n");
+    int err = pthread_create(&t, &attr, INIT_AFL_MUTATION_SESSION, process);
+    if (err != 0) {
+      printf("[!] %d: Unable to CREATE AFL++ mutation thread.\n", err);
+      printf("[!] Tearing down parent process gracefully...\n");
+      destroy_and_exit_process(process, err);
+    }
 }
+// AFL++ EOS
 
 #ifdef HFND_FUZZING_ENTRY_FUNCTION
  HFND_FUZZING_ENTRY_FUNCTION(int argc, const char *const *argv)
@@ -491,13 +655,23 @@
  int main(int argc, const char *const *argv)
 #endif
 {
+
+    // AFL++ SNIP
+    process_rec *process;
+    if (getenv("CRASH_TEST") == NULL) {
+      INIT_AFL_THREAD(process);
+    } else {
+      printf("[!] Crash Test Mode Initialized:\n");
+    }
+    // AFL++ EOS
+
     char c;
     int showcompile = 0, showdirectives = 0;
     const char *confname = SERVER_CONFIG_FILE;
     const char *def_server_root = HTTPD_ROOT;
     const char *temp_error_log = NULL;
     const char *error;
-    process_rec *process;
+
     apr_pool_t *pconf;
     apr_pool_t *plog; /* Pool of log streams, reset _after_ each read of conf */
     apr_pool_t *ptemp; /* Pool for temporary config stuff, reset often */
@@ -750,6 +924,9 @@
         int exit_status;
 
         if (signal_server(&exit_status, pconf) != 0) {
+            // AFL++ SNIP
+            printf("[!] ERROR: Server signaled out!\n");
+            // AFL++ EOS
             destroy_and_exit_process(process, exit_status);
         }
     }
