--- server/main.c	2020-09-02 09:24:31.000000000 -0600
+++ server/main.AFL++.c	2020-09-08 09:57:18.000000000 -0600
@@ -14,6 +14,44 @@
  * limitations under the License.
  */
 
+// AFL++ SNIP
+#include <arpa/inet.h>
+#include <errno.h>
+#include <linux/sched.h>
+#include <net/if.h>
+#include <net/route.h>
+#include <netinet/ip6.h>
+#include <netinet/tcp.h>
+#include <sched.h>
+#include <signal.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h>
+#include <sys/ioctl.h>
+#include <sys/resource.h>
+#include <sys/socket.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+/*
+#ifndef __AFL_FUZZ_TESTCASE_LEN
+  ssize_t fuzz_len;
+  #define __AFL_FUZZ_TESTCASE_LEN fuzz_len
+  unsigned char fuzz_buf[1024000];
+  #define __AFL_FUZZ_TESTCASE_BUF fuzz_buf
+  #define __AFL_FUZZ_INIT() void sync(void);
+  #define __AFL_LOOP(x) ((fuzz_len = read(0, fuzz_buf, sizeof(fuzz_buf))) > 0 ?
+  #define __AFL_INIT() sync()
+#endif
+
+__AFL_FUZZ_INIT();
+*/
+// AFL++ EOS
+
 #include "apr.h"
 #include "apr_strings.h"
 #include "apr_getopt.h"
@@ -482,22 +520,130 @@
     ap_log_error(APLOG_MARK, APLOG_STARTUP, 0, NULL,
                  "  -X                 : debug mode (only one worker, do not detach)");
 
-    destroy_and_exit_process(process, 1);
+// AFL++ SNIP
+    destroy_and_exit_process(process, 0);
+// AFL++ EOS
 }
 
+// AFL++ SNIP
+static void INIT_AFL_MUTATION_SESSION(process_rec *process)
+{
+  int BUFSIZE=1024*1024;
+  usleep(10000);
+  char buf[BUFSIZE+1];
+  while ( __AFL_LOOP(30000) ){
+    printf("[+] Looping\n");
+    memset(buf, 0, BUFSIZE);
+    size_t read_bytes = read(0, buf, BUFSIZE);
+    buf[BUFSIZE-2] = '\r';
+    buf[BUFSIZE-1] = '\n';
+    buf[BUFSIZE] = '\0';
+
+    int sockfd = socket(AF_INET, SOCK_STREAM, IPPROTO_IP);
+    if (sockfd == -1) {
+      perror("socket");
+      _exit(1);
+    }
+
+    int sz = (1024 * 1024);
+    if (setsockopt(sockfd, SOL_SOCKET, SO_SNDBUF, &sz, sizeof(sz)) == -1) {
+      perror("setsockopt");
+      exit(1);
+    }
+
+    printf("[+] Connecting\n", buf);
+
+    struct sockaddr_in saddr;
+    saddr.sin_family = AF_INET;
+    saddr.sin_port = htons(80);
+    saddr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+    if (connect(sockfd, &saddr, sizeof(saddr)) == -1) {
+      printf("[-] Connect failed\n");
+      perror("connect");
+      continue;
+    }
+
+    printf("[+] Sending buf %s\n", buf);
+
+    if (send(sockfd, buf, read_bytes, MSG_NOSIGNAL) != read_bytes) {
+      perror("send() failed 1");
+      exit(1);
+    }
+
+    printf("[+] Buf sent %s\n", &buf);
+
+    if (shutdown(sockfd, SHUT_WR) == -1) {
+      perror("shutdown");
+      exit(1);
+    }
+
+    char b[1024 * 1024];
+    while (recv(sockfd, b, sizeof(b), MSG_WAITALL) > 0) {
+      printf("[+] Received %s\n", b);
+    }
+
+    close(sockfd);
+    printf("[+] Nice run\n");
+  }
+  printf("[+] Whew lad!\n");
+  usleep(100000);
+
+  int rem_pidlog;
+  printf("Remove DEFAULT_PIDLOG\n");
+  rem_pidlog = remove("/fuzz_session/httpd/logs/httpd.pid");
+  printf("rem_pidlog response: %d", rem_pidlog);
+
+  if ( rem_pidlog != 0 ) {
+    printf("[!] Ugh, unable to remove %s", DEFAULT_PIDLOG);
+    _exit(1);
+  } else {
+    printf("[+] REMOVED %s\n", DEFAULT_PIDLOG);
+    _exit(0);
+  }
+}
+
+static void INIT_AFL_THREAD(process_rec *process)
+{
+    pthread_t t;
+    pthread_attr_t attr;
+
+    pthread_attr_init(&attr);
+    pthread_attr_setstacksize(&attr, 1024 * 1024 * 8);
+    pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);
+
+    printf("[*]: CREATING AFL++ mutation thread...\n");
+    int err = pthread_create(&t, &attr, INIT_AFL_MUTATION_SESSION, process);
+    if (err != 0) {
+      printf("[!] %d: Unable to CREATE AFL++ mutation thread.\n", err);
+      printf("[!] Tearing down parent process gracefully...\n");
+      destroy_and_exit_process(process, err);
+    }
+}
+// AFL++ EOS
+
+
 #ifdef HFND_FUZZING_ENTRY_FUNCTION
  HFND_FUZZING_ENTRY_FUNCTION(int argc, const char *const *argv)
 #else
  int main(int argc, const char *const *argv)
 #endif
 {
+    // AFL++ SNIP
+    process_rec *process;
+    if (getenv("CRASH_TEST") == NULL) {
+      INIT_AFL_THREAD(process);
+    } else {
+      printf("[!] Crash Test Mode Initialized:\n");
+    }
+    // AFL++ EOS
+
     char c;
     int showcompile = 0, showdirectives = 0;
     const char *confname = SERVER_CONFIG_FILE;
     const char *def_server_root = HTTPD_ROOT;
     const char *temp_error_log = NULL;
     const char *error;
-    process_rec *process;
+
     apr_pool_t *pconf;
     apr_pool_t *plog; /* Pool of log streams, reset _after_ each read of conf */
     apr_pool_t *ptemp; /* Pool for temporary config stuff, reset often */
@@ -750,6 +896,9 @@
         int exit_status;
 
         if (signal_server(&exit_status, pconf) != 0) {
+            // AFL++ SNIP
+            printf("[!] ERROR: Server signaled out!\n");
+            // AFL++ EOS
             destroy_and_exit_process(process, exit_status);
         }
     }
